### ISSUE TITLE
Join two JinqStreams

### ISSUE THREAD

**Issue by oberien** [my2iu/Jinq#12]
I have two queries, which are both working and translatable into sql. The elements streamed in these queries are different and not connected. If I have two pre-filtered JPAJinqStreams and want to join them on one property, Jinq tells me, it is not possible, even though in SQL it is one of the easiest things to do.

Enough on these abstract stuff, let's get to an example:
I have two classes `Class1` and `Class2`, both of which can be tagged with class `Tag`. I don't want `List<Tag>` to be a delegate of all classes being able to be tagged (in this example `Class1` and `Class2`).
To get all valid Tag-Class1-Pairs, this query works perfectly as expected:

``` java
streams.streamAll(em, Class1.class)
    .join((c, source) -> source.stream(Tag.class))
    .where(p -> p.getOne().getId() == p.getTwo().getTaggedId())
    .forEach(System.out::println);
```

Producing:

``` sql
SELECT A, B FROM Class1 A, Tag B WHERE A.id = B.taggedId
```

Now I'd like to pre-filter both streams (e.g. permission-wise) and get them separately by calling Service-Functions:

``` java
JPAJinqStream<Class1> classStream = service.getFilteredClassStream();
JPAJinqStream<Tag> tagStream = service.getFilteredTagStream();

classStream.join(c -> tagStream)
    .where(p -> p.getOne().getId() == p.getTwo().getTaggedId())
    .forEach(System.out::println);
```

This for whatever reason results in `java.lang.reflect.InvocationTargetException Caused by: java.lang.IllegalArgumentException: Could not extract code from lambda. This error sometimes occurs because your lambda references objects that aren't Serializable.`

But actually there is no major difference semantic-wise. In the first example, I create the new stream from inside the stream. In the second example I use one from outside. In the first example `source.stream` produces `JinqStream<U>`, which is the same as `getFilteredTagStream()` returning `JPAJinqStream<Tag> extends JinqStream<Tag>`.

So my question is: How do I join two different streams, produced in different contexts?

**Comment by my2iu** [comment]
This is an annoying restriction that happens because of limitations of Java 8 lambdas. Jinq is only able to analyze lambdas if the lambdas are serializable. Lambdas are only serializable if all the external variables accessed by the lambda are serializable. Unfortunately, in

```
classStream.join( c -> tagStream )
```

the `tagStream` is not serializable. I could maybe hack things to support this particular way of expressing things, but it makes things inconsistent (e.g. I could make JinqStreams serializable even though they aren't) and there are workarounds, so I'm reluctant to do that.

Here are some notes about workarounds. (Note: I haven't tested the code I'm showing in the examples. These are just some things I'm writing from memory):

1. It's possible to express the queries, but you need to assemble the whole query all at once like a normal SQL query.

```
classStream.join((c, source) -> 
     source.stream(Tag.class).where(t -> t.getKey().equals("name")))
```

2. You can also post-filter the queries instead of pre-filtering them

```
{
   JinqStream<Pair<A, Tag>> stream = classStream.join((c, source) -> source.stream(Tag.class));
   stream = filterClassTag(stream, "name");
}

<A> JinqStream<Pair<A, Tag>> filterClassTag(JinqStream<Pair<A, Tag>> stream, String tagName)
{
   return stream.where(s -> s.getSecond().getKey().equals(tagName));
}
```

3. Just as a side note, if you find you're using `join`, it usually means you can just add the relation to your entities, which sometimes results in cleaner code. In this case, I don't think it helps that much.

One possibility:
```
streams.streamAll(em, Class1.class)
    .join(c -> c.getTags())
```

Another possibility with filtering but no join:
```
classStream.where(c -> 
     JinqStream.from(c.getTags()).where(t -> t.getKey().equals("name")).count() > 0);
```

**Comment by alahijani** [comment]
I'm not sure if I agree with the evaluation that making JinqStreams serializable is just a "hack". I makes real sense to me, as philosophically, translating a JinqStream to a query is by itself a form of serialization.

When `tagStream.writeObject` is called, it should translate the query and serialize the translation result (with any parameters) to the `ObjectOutputStream`. The serialized representation has no link to an EntityManager or a SQL connection, so you cannot stream it directly, but upon deserialization it can be used to create an instance of JPQLQuery which then can be used recursively by other JinqStreams to translate bigger queries like the above example.

As a bonus, this will also open the door for more complex nested queries like a `JinqStream.contains(T)` method that is intrinsically supported by the lambda analyser and can be translated to queries of the form `WHERE ? IN (SELECT ...)`.

**Comment by my2iu** [comment]
I spent a while looking at this issue a year and a half ago. I actually wrote some code for it too. It's possible to get it to work, but it's inelegant. The changes make it possible to work around problems with Java 8 lambdas, but the trade-off is that it suddenly becomes possible to write all sorts of code in Jinq that doesn't work.

In general, streams are not serializable. They are entirely transient state. You can only use streams once. It's not clear what it means when you serialize and then deserialize a stream. Does the original stream work any more? Does the new stream actually work? What's the point of serializing a stream if you can't create a new stream by deserializing it? What happens if you deserialize a stream on a different computer? `JinqStream`s are weirder because they contain an `EntityManager`, so they are supposed to only exist within a database transaction. What happens when you join streams from different transactions? When you deserialize a `JinqStream`, will it create a new database transaction or try to reattach itself to the original transaction?

If there wasn't a workaround, I would definitely have added support for doing that. But there are workarounds, so I thought it was best to keep the API cleaner.

As a side note, Jinq does support complex nested queries. Queries like `WHERE ? IN (SELECT ...)` should be supported through the use of `JPQL.isIn()` as well as some other mechanisms.

**Comment by Blackhole16** [comment]
Image the following scenario:

You have multiple layers of Abstration (which is the default for most complex SaaS):
* Data Access Layer
* Service Layer
* API

The Services are fetching / filtering the data, inside of the API they are wrapped and given to the client. You now have the Services return JinqStreams instead of List<Class>, so that the API can also mix up different Data without performance loss, as everything is translated into SQL.

Regarding my example none of the workarounds is actually working, as different Services are returning their own streams. One returns pre-filtered Tags and another one Taggable-Items. Combining these Streams must be done by manually joining them in Java, which is just very slow and bad, also taking into account, that actually two separate DB-transactions are done. Combining everything into one stream being translated into one transaction / one sql-query could reduce load times by huge amounts.

For that Use-Case unfortunately there is no other solution then to accept external JinqStreams and merge them into other JinqStreams.

Actually I also don't really see a major difference between
```java
stream.join((e, source) -> source.stream(Clazz))
```
and
```java
externalStream = ...
stream.join(e -> externalStream)
```
other than that one is generated in `stream` and the other one is generated outside of `stream`. For convenience actually there shouldn't be made a difference from this library's point of view, as it improves usage-convenience a lot.

In conclusion this use case shouldn't be too uncommon and implementing this kind of functionality improves usage-convenience a lot.

**Comment by my2iu** [comment]
The difference between your two examples is that one of them is supported by Java 8 lambda serialization and one of them isn't. Unfortunately, Jinq requires Java 8 lambda serialization to work.

Although I agree that it would be really nice if Jinq supported queries like that, it can't do that without making things really ugly. The workaround is to fix your data abstractions. If application code is allowed to make queries, then you don't need all those service layers and data access layers. If application code isn't allowed to make queries, then you should move the join code up into the data access layers instead of in the application code. Having all these data and service layers while still joining things and creating queries in the application layer suggests that there is a problem with your abstractions. This problem would exist regardless of which query system you are using.

**Comment by Blackhole16** [comment]
Yep you're right regarding the remodeling of the abstraction layers. But there is still another use-case ;) It's called code-duplication.

If I have `getFooStream`, `getTagStream` and `getFooWithTagsStream`, then I need to duplicate at least one of the two basic methods, resulting in harder maintainability.

```java
public JinqStream<Foo> getFooStream() {
   [heavy filtering]
}

public JinqStream<Tag> getTagStream() {
  [heavy filtering]
}

// preferred
public JinqStream<Pair<Foo, Tag>> getFooWithTagsStream() {
  getFooStream.join(f -> getTagStream()).where(p -> p.getOne().getId() == p.getTwo().getTaggedId())
}

// actual
public JinqStream<Pair<Foo, Tag>> getFooWithTagsStream() {
  getFooStream.join((f, source) -> [heavy filtering of Tag, completely equal to getTagStream()]).where(p -> p.getOne().getId() == p.getTwo().getTaggedId())
}
```

I'm not saying, that this must be implemented, but I'd like to show, that there are use-cases of this other than wrong data structure :) However if this is too hard / hacky to implement (yet), it's fine as this library IMO already fills a huge gap in Java.

If this is not implemented yet, could this issue be marked as RFC and reopened? Because having this functionality would be pretty nifty.

PS: This cross-joining is actually possible in C#'s Linq. Regarding to @main--
```c#
from foo in foos
join tag in tags where tag.TaggedId equals foo.Id
```

**Comment by my2iu** [comment]
Sure, I'll leave the issue in the open state. The feature needed by Jinq to implement the feature properly is for the Java reflection APIs to support lambdas. I've been pressing the Java people to add it in. They were very reluctant previously, but they might be becoming more receptive to the idea. But it's very low on their priority list, unfortunately.

**Comment by Blackhole16** [comment]
For the time being, is there some kind of (smelly) workaround that I could use on my side to make this work? I found several other use-cases (e.g. joining two independent objects on some id, which are not suitable to remodeling) and this is actually keeping me from using this library, which I definitely would like to use.

**Comment by my2iu** [comment]
I keep feeling like there's some larger architectural issue that's missing here because this problem occurs with almost all Java database query systems, and people are ok with it. You may be focusing too much on getting this particular solution to work when if you stepped back and looked at the bigger picture, there's an alternate architecture that makes everything easy. I would have to look over your architecture to understand what's going on though.

I suggested a workaround earlier. Instead of filtering things in `getFilteredTagStream()`, you can join and filter things afterwards.

So you would have something like this:

```
JinqStream<Foo> foos = getFooStream();
JinqStream<Foo, Tag> foosWithTags = joinTags(foos);
JinqStream<Foo, Tag> foosWithTagsFiltered = filterJoinedTags(foosWithTags);
```

I don't see how that's much different from your desired architecture of

```
JinqStream<Foo> foos = getFooStream();
JinqStream<Tag> tags = getTags();
JinqStream<Tag> tagsFiltered = filterTags(tags);
JinqStream<Foo, Tag> foosWithTagsFiltered = joinTags(foos, tagsFiltered);
```

It's pretty much the exact same code except the ordering is a little different. Filtering before or after joining makes no difference. You get the same performance. The query generated by Jinq in both cases would probably be exactly the same. The `filterJoinedTags()` method can be made generic as I showed in the first reply, so you can reuse the code for things other than `Foo`. I don't understand why you absolutely must filter the tags before joining them instead of after.

**Comment by my2iu** [comment]
Actually, I now remember some additional issues I encountered when I looked into this problem last year. I'm just including it here as documentation for myself, so please don't feel obligated to read it.

One of the main issues I was encountering was that the API was not "executable" in Java. If you were to run the code using in-memory data and normal streams, the code would cause an exception. This is problematic because the philosophy of the API is that it's theoretically possible to run all the same operations on regular Lists and Collections, and Jinq is simply translating everything into database queries behind the scenes.

So let's say you have:

```
List<Foo> fooList = ...;
List<Tag> tagList = ...;

JinqStream<Foo> foos = (JinqStream<Foo>)fooList.stream();
JinqStream<Tag> tags = (JinqStream<Tag>)tagList.stream();

foos.join( foo -> tags );
```

Running the code will generate this exception:

```
java.lang.IllegalStateException: 
  stream has already been operated upon or closed
```

The problem is that the code for join would theoretically look something like this:

```
class JinqStream<T> {
  <U> JinqStream<Pair<T,U>> join(Function<T,JinqStream<U>> fun)
  {
    joined = new List
    foreach t:
      JinqStream<U> toJoin = fun.apply(t)
      foreach u in toJoin:
        joined.add(Pair(t, u))
    return joined.stream()
  }
}
```

So during the processing of the join, the same stream is scanned multiple times, causing an exception to be thrown. The function passed to `join()` is supposed to create a new stream every time it is called, not return the same stream again and again.

If things are translated into queries, all this isn't a problem because the "equivalent" database query doesn't actually involve processing streams in this way. But philosophically, there is a problem that Jinq is no longer trying to emulate Java behaviour with a database query. The actual Java code is incorrect and has errors in it, but Jinq is still generating running code for it that "works." This opens up a can of worms that I couldn't resolve.

One possible alternative is to create a different API like

```
<U> JinqStream<Pair<T,U>> crossJoin(JinqStream<U> stream)
```

This is potentially problematic in that it's much less powerful than normal joins and it's hard to explain exactly why it's needed compared to the normal join. It's not clear if this is "wrong" in that it allows people to do things that they shouldn't be doing. Also, it requires a lot of checking to see if entity managers match etc.

**Comment by my2iu** [comment]
I have now implemented the crossJoin() method for directly joining two streams in this way.
